src/main/scala/PropertyChecker.scala

```diff
@@ -402,4 +412,89 @@ object PropertyChecker {
        (objectNode, warnings, csvwPropertyType)
     }
   }
+
+  def tableSchemaProperty(csvwPropertyType: PropertyType.Value): (JsonNode, String, String) => (JsonNode, Array[String], PropertyType.Value) = {
+    def tableSchemaPropertyInternal(value: JsonNode, baseUrl: String, lang: String): (JsonNode, Array[String], PropertyType.Value) = {
+      var schemaBaseUrl = new URL(baseUrl)
+      var schemaLang = lang
+
+      var schemaJson: ObjectNode = null
+      if(value.isTextual) {
+        val schemaUrl = new URL(new URL(baseUrl), value.asText())
+        schemaJson = PropertyChecker.mapper.readTree(schemaUrl).asInstanceOf[ObjectNode]
+        if (!schemaJson.path("@id").isMissingNode) {
+          // Do something here as object node put method doesn't allow uri object
+          val absoluteSchemaUrl = new URL(schemaUrl, schemaJson.get("@id").asText())
+          schemaJson.put("@id", absoluteSchemaUrl.toString)
+        } else {
+          schemaJson.put("@id", schemaUrl.toString)
+        }
+        val (newSchemaBaseUrl, newSchemaLang) = fetchSchemaBaseUrlAndLangAndRemoveContext(schemaJson, schemaBaseUrl, schemaLang)
+        schemaBaseUrl = newSchemaBaseUrl
+        schemaLang = newSchemaLang
+      } else if(value.isObject) {
+        schemaJson = value.deepCopy()
+      } else {
+        return (NullNode.getInstance(), Array[String](PropertyChecker.invalidValueWarning), PropertyType.Table)
+      }
+
+      var warnings = Array[String]()
+      val fieldsAndValues = Array.from(schemaJson.fields.asScala)
+      for(fieldAndValue <- fieldsAndValues) {
+        warnings = validateObjectAndUpdateSchemaJson(schemaJson, schemaBaseUrl, schemaLang, fieldAndValue.getKey, fieldAndValue.getValue)
+      }
+      return (schemaJson, warnings, PropertyType.Table)
+    }
+    return tableSchemaPropertyInternal
+  }
+
+  def validateObjectAndUpdateSchemaJson(schemaJson: ObjectNode, schemaBaseUrl: URL, schemaLang: String,
+                                        property: String, value: JsonNode):(Array[String]) = {
+    var warnings = Array[String]()
+    if(property == "@id") {
+      val matcher = PropertyChecker.startsWithUnderscore.pattern.matcher(value.asText())
+      if (matcher.matches) {
+        throw new MetadataError(s"@id ${value.asText} starts with _:")
+      }
+    } else if(property == "@type") {
+      if (value.asText() != "Schema") {
+        throw new MetadataError("@type of schema is not 'Schema'")
+      }
+    } else {
+      val (validatedV, warningsForP, propertyType) = checkProperty(property, value, schemaBaseUrl.toString, schemaLang)
+      if((propertyType == PropertyType.Schema || propertyType == PropertyType.Inherited) && warningsForP.isEmpty) {
+        schemaJson.set(property, validatedV)
+      } else {
+        schemaJson.remove(property)
+        if(propertyType != PropertyType.Schema || propertyType != PropertyType.Inherited) {
```

Comment:
I suspect this should be `propertyType != PropertyType.Schema && propertyType != PropertyType.Inherited`.

Reply:
Technically `propertyType != PropertyType.Schema || propertyType != PropertyType.Inherited` will always be true. If it's `PropertyType.Schema` then it can't also be `PropertyType.Inherited` (and vice versa).

Reply:
Got it. This mistake happened when I converted the `unless` from ruby to scala.